test/unit/data-type.js

```diff
@@ -75,9 +75,322 @@ describe('Data Types', function() {
       const parameter = { value: value, scale: scale };
       parameter.value.nanosecondDelta = nanosecondDelta;
 
-      const buffer = new WritableTrackingBuffer(16);
-      type.writeParameterData(buffer, parameter, { useUTC: false }, () => {});
+      const buffer = new WritableTrackingBuffer(0);
+      type.writeParameterData(buffer, parameter, { useUTC: false }, () => { });
       assert.deepEqual(buffer.data, expectedBuffer);
     }
   });
+
+  it('should writeParameterData bigInt', () => {
+    const type = TYPES.typeByName.BigInt;
+
+    for (const [value, expected] of [
+      [123456789, Buffer.from('0815cd5b0700000000', 'hex')],
+      [null, Buffer.from('00', 'hex')],
+    ]) {
+      const buffer = new WritableTrackingBuffer(0);
+      const parameterValue = { value };
+      type.writeParameterData(buffer, parameterValue, { useUTC: false }, () => { });
+      assert.isTrue(buffer.data.equals(expected));
+    }
+  });
+
+  it('should writeParameterData binary', () => {
+    const type = TYPES.typeByName.Binary;
+
+    for (const [value, expected] of [
+      [Buffer.from([0x12, 0x34, 0x00, 0x00]), Buffer.from('040012340000', 'hex')],
+      [null, Buffer.from('ffff', 'hex')]
+    ]) {
+      const buffer = new WritableTrackingBuffer(0);
+      const parameterValue = { value, length: 4 };
+      type.writeParameterData(buffer, parameterValue, { useUTC: false }, () => { });
+      assert.isTrue(buffer.data.equals(expected));
+    }
+  });
```

Comment:
All these test cases that call `writeParameterData` are written in a synchronous fashion, while the `writeParameterData` is executing in a potentially asynchronous fashion - e.g. for `TVP`, the passed in callback is sometimes called after one or multiple `setImmediate` calls.

The reason why these test cases are written in a synchronous fashion is that initially, `writeParameterData` was a regular synchronous function and did not take a callback parameter, and when that was changed, these test cases were not updated correctly.

A better way to write these test cases would be something like this:

```js
  it('should writeParameterData binary (Buffer)', function(done) {
    const value = Buffer.from([0x12, 0x34, 0x00, 0x00]);
    const expected = Buffer.from('040012340000', 'hex');

    const type = TYPES.typeByName.Binary;

    const buffer = new WritableTrackingBuffer(0);
    const parameterValue = { value, length: 4 };
    type.writeParameterData(buffer, parameterValue, { useUTC: false }, () => {
      assert.isTrue(buffer.data.equals(expected));

      done();
    });
  });

  it('should writeParameterData binary (null)', function(done) {
    const value = null;
    const expected = Buffer.from('ffff', 'hex');

    const type = TYPES.typeByName.Binary;

    const buffer = new WritableTrackingBuffer(0);
    const parameterValue = { value, length: 4 };

    type.writeParameterData(buffer, parameterValue, { useUTC: false }, () => {
      assert.isTrue(buffer.data.equals(expected));

      done();
    });
  });
```

Notice the addition of `done` to the test function, and calling `done` inside the callback passed to `writeParameterData`.

Also, the assertion was moved into the callback to `writeParameterData` as well.

I also had to break the test into two separate test cases, because testing multiple input/output combinations inside an asynchronous test function is awkward, and it's easier to just create multiple test cases.